runner/runners/queue.go

```diff
@@ -123,16 +146,18 @@ func NewQueueRunner(
 }
 
 func NewSingleRunner(
-	exec execer.Execer, filer snapshot.Filer, idc snapshot.InitDoneCh, output runner.OutputCreator, tmp *temp.TempDir, stat stats.StatsReceiver) runner.Service {
-	return NewQueueRunner(exec, filer, idc, output, tmp, 0, stat)
+	exec execer.Execer, filerMap runner.RunTypeMap, output runner.OutputCreator, tmp *temp.TempDir, stat stats.StatsReceiver) runner.Service {
+	return NewQueueRunner(exec, filerMap, output, tmp, 0, stat)
 }
 
 // QueueController maintains a queue of commands to run (up to capacity).
 // Manages updates to underlying Filer via Filer's Update interface,
 // if a non-zero update interval is defined (updates and tasks cannot run concurrently)
 type QueueController struct {
 	inv           *Invoker
-	filer         snapshot.Filer
+	filerMap      runner.RunTypeMap
+	updateReq     map[runner.RunType]bool
+	updateLock    sync.Mutex
```

Comment:
All this locking stuff seems unwieldy, I can't easily tell if it's necessary. It seems like you can just use yet another channel and send it the run type every time the respective ticker updates.

Reply:
I'm using the lock only when tickers update the map of bool flags that indicate what filers have update requests, and then once in the loop to read & reset the flags.  To queue these up on a channel it would have to be buffered -- to what length? how do updateTickers know if they have a request sitting on the channel or not without draining it? didn't see a clear way of doing that

Reply:
Why does it have to be buffered? They end up being processed serially right? Seems like we can do a blocking send and ultimately it doesn't matter if it gets backed up or not, although I'd expect it shouldn't with reasonable ticker times.

Reply:
I think an updater could get starved out in this case, if the request channel has a request and before it's drained by the updater (some task is running and taking a while), several updaters hit their tick again and compete to make the next request. Possible with sane values

Reply:
True. In that case I'd use a single ticker with the gcd of filer intervals and then on each tick loop through the filers seeing who all deserves a turn.

Reply:
Up to  you though, whatever you think will be the most understandable.